lib/appoptics_apm/inst/grpc_client.rb

```diff
@@ -0,0 +1,161 @@
+        block_given? ? response.each { |r| yield r } : response
+      rescue => e
+        unless e.instance_variable_get(:@exn_logged)
+          context_from_incoming
+          AppOpticsAPM::API.log_exception('grpc_client', e)
+          AppOpticsAPM::API.log_exit('grpc_client', exit_tags(@tags))
+        end
+        raise e
+      end
+
+      private
+
+      def unary_response(req, type: , metadata: , without:)
+        tags = grpc_tags(type, metadata[:method] || metadata_to_send[:method])
+        AppOpticsAPM::SDK.trace('grpc_client', tags) do
+          metadata['x-trace'] = AppOpticsAPM::Context.toString
+          AppOpticsAPM::SDK.set_transaction_name(metadata[:method]) if AppOpticsAPM.transaction_name.nil?
+          begin
+            send(without, req, metadata: metadata)
+          ensure
+            exit_tags(tags)
+            context_from_incoming
+          end
+        end
+      end
+
+      def patch_receive_and_check_status
+        def self.receive_and_check_status # need to patch this so that log_exit can be called after the enum is consumed
+          super
+          context_from_incoming
+        rescue => e
+          context_from_incoming
+          AppOpticsAPM::API.log_exception('grpc_client', e)
+          raise e
+        ensure
+          AppOpticsAPM::API.log_exit('grpc_client', exit_tags(@tags))
+        end
+      end
+
+      def patch_set_input_stream_done
+        # need to patch this instance method so that log_exit can be called after the enum is consumed
+        def self.set_input_stream_done
+          return if status.nil?
+          context_from_incoming
+          if status.code > 0
+            AppOpticsAPM::API.log_exception('grpc_client', $!)
+          end
+          AppOpticsAPM::API.log_exit('grpc_client', exit_tags(@tags))
+          super
+        end
+      end
+
+      def context_from_incoming
+        xtrace ||= @call.trailing_metadata['x-trace'] if @call.trailing_metadata && @call.trailing_metadata['x-trace']
+        xtrace ||= @call.metadata['x-trace'] if @call.metadata && @call.metadata['x-trace']
+        AppOpticsAPM::Context.fromString(xtrace) if xtrace
+      end
+
+      def exit_tags(tags)
+        # we need to translate the status.code, it is not the status.details we want, they are not matching 1:1
+        tags['GRPCStatus'] ||= @call.status ? StatusCodes[@call.status.code].to_s : 'UNKNOWN'
+        tags.delete('Backtrace')
+        tags
+      end
+    end
+
+  end
+end
+
+if defined?(::GRPC) && AppOpticsAPM::Config[:grpc_client][:enabled]
+  AppOpticsAPM.logger.info '[appoptics_apm/loading] Instrumenting GRPC' if AppOpticsAPM::Config[:verbose]
+
+  # Client side is instrumented in ActiveCall and ClientStub
+  ::AppOpticsAPM::Util.send_include(::GRPC::ActiveCall, ::AppOpticsAPM::GRPC::ActiveCall)
+
+  GRPC_ClientStub_ops = [:request_response, :client_streamer, :server_streamer, :bidi_streamer]
+  module GRPC
+    class ClientStub
```

Comment:
Is this where the GRPC methods get wrapped so that they can be traced? It seems like this is defining methods. That part kind of makes sense to me. I don't quite get the `send` - from reading it looks like there is no object that `send` is operating on so that means the global Object is used. Not sure how that works here. Not even sure I'm right about what's going on here.

Reply:
`send` is a way to call a method when you have a method name as a string.
Here it is called from an instance method that is getting defined with `define_method` (used to define a method from a string). So it calls the method on the current instance.

Reply:
i don't understand alias but i did understand (i think) that you were calling the original, unmodified method, with `send`. this was an example of meta-programming, i.e., writing methods on the fly?